mono/metadata/threads.c

```diff
@@ -562,10 +562,16 @@ create_internal_thread_object (void)
 
 	thread->apartment_state = ThreadApartmentState_Unknown;
 	thread->managed_id = get_next_managed_thread_id ();
+	/* 
+	 * Boehm incremental is not actually "moving", it does not need the thread_pinning_ref.
+	 * But having it causes problems when unregistering the root after domain reload.
+	 */
+#if !defined(HAVE_BOEHM_GC) 
```

Comment:
We cannot do GC specific checks in standard `metadata` code since it's only built once for both GCs. See 

https://github.com/Unity-Technologies/mono/blob/unity-master/mono/metadata/Makefile.am#L271 

and

https://github.com/Unity-Technologies/mono/blob/unity-master/mono/metadata/Makefile.am#L318

We can either move this file to the GC dependent list or something add a runtime check to determine if we are using Boehm.

Reply:
Oh. In that case, I guess I new fix is in order for https://github.com/Unity-Technologies/mono/commit/85e0af093381f1404c4b422b54c8bff9fe8239ec as well. I'll think of something.

Possibly I should not set mono_gc_is_moving incremental boehm (since it is not moving), but have a mono_gc_needs_write_barriers instead - but then it would be harder to sync with upstream.

Reply:
I could be wrong. If experimentation indicates things are built twice I am fine with the changes.